### PR TITLE
LargeFileUpload task 

### PR DESCRIPTION
**Pull request by andrueastman**
### Changes proposed in this pull request
- Builds upon fix for extra memory allocation when doing chunked upload( from PR #529 credit to @zsybupt  )
- Wraps the ChunkedUploadProvider in a LargeFileUpload task
- Adds support for providing a callback mechanism for monitoring uploads

This work is done in accordance with the spec found at this link. https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/FileUploadTask.md

Essentially, one may perform a large upload by following the steps below.

1. create an upload session
```cs
// Create upload session 
// POST /v1.0/drive/items/01KGPRHTV6Y2GOVW7725BZO354PWSELRRZ:/SWEBOKv3.pdf:/microsoft.graph.createUploadSession
var uploadSession = await graphClient.Drive.Items[itemId].ItemWithPath("SWEBOK.pdf").CreateUploadSession().Request().PostAsync();
```

2. Create the task using the session.
```cs
// Create task
var maxChunkSize = 320 * 1024; // 320 KB - Change this to your chunk size. 5MB is the default.
var largeFileUpload = new LargeFileUpload(uploadSession, graphClient, stream, maxChunkSize);
```

3. Create an upload monitor
```cs
    public class MyProgress : IProgressCallback
    {
        public void OnFailure(ClientException clientException)
        {
            Console.WriteLine(clientException.Message);
        }

        public void OnSuccess(DriveItem result)
        {
            Console.WriteLine("Download completed with id below");
            Console.WriteLine(result.Id);
        }

        public void UpdateProgress(long current, long max)
        {
            long percentage = (current * 100) / max ;
            Console.WriteLine("Upload in progress. " + current + " bytes of " + max + "bytes. " + percentage + " percent complete");
        }
    }
```

4. Upload the file 
```cs
uploadedFile = await largeFileUpload.ResumeAsync(new MyProgress());
```

**Alternatively**, one may choose to do the file upload in slices by themselves by replacing step 3 and 4 with

```cs
// Setup the chunk request necessities
var slicesRequests = largeFileUpload.GetUploadSlicesRequests();
var trackedExceptions = new List<Exception>();
DriveItem itemResult = null;

//upload the chunks
foreach (var request in slicesRequests)
{
    // Send chunk request
    var result = await largeFileUpload.UploadSliceAsync(request, trackedExceptions);
    // Do your updates here: update progress bar, etc.
    Console.WriteLine($"File uploading in progress. {request.RangeEnd} of {stream.Length} bytes uploaded");

    if (result.UploadSucceeded)
    {
        itemResult = result.ItemResponse;
        Console.WriteLine($"File uploading complete");
    }
}
```
#### Notes
One can find a sample app built upon this PR(using this repo as a git submodule) to showcase and test the above usage at the following link.
https://github.com/andrueastman/UploadFileSample

### CODE REVIEW

File: src/Microsoft.Graph/Tasks/IProgressCallBack.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph
+{
+    /// <summary>
+    /// Interface for to enbale handling of progress/errors for a long upload process.
```
[the remark this answers is not in the document]
Comment (darrelmiller) [reply src/Microsoft.Graph/Tasks/IProgressCallBack.cs]: .NET has an interface built in https://docs.microsoft.com/en-us/dotnet/api/system.iprogress-1?view=netframework-4.8

**Comment by darrelmiller** [comment]
99% of this code should be in Core.   We need to make this generic so that it is not tied to DriveItem.  We will want to use this code for large attachment uploads also.

**Comment by darrelmiller** [comment]
We cannot use the term Chunk anywhere in our interface or samples.   This is not chunking.

File: src/Microsoft.Graph/Requests/Helpers/ChunkedUploadProvider.cs [hunk 2]
```diff
@@ -147,33 +149,54 @@ public async Task<DriveItem> UploadAsync(int maxTries = 3, IEnumerable<Option> o
                 if (uploadTries < maxTries)
```
Comment (peombwa) [inline src/Microsoft.Graph/Requests/Helpers/ChunkedUploadProvider.cs]: This retry logic will result in 3^3 retries (by default) for 429, 503 and 504 HTTP responses when the service doesn't return a successful result. The underlying HTTP client already has a RetryHandler that handles retries for 429, 503 and 504 responses.

**Comment by Misiu** [comment]
> 99% of this code should be in Core. We need to make this generic so that it is not tied to DriveItem. We will want to use this code for large attachment uploads also.

I'm waiting for a large file upload while sending emails. It was announced some time ago (https://docs.microsoft.com/en-us/graph/outlook-large-attachments?tabs=http), but its still unsupported here and in beta lib.

**Comment by andrueastman** [comment]
Thanks for all the feedback on this PR. 
I am going to close this PR and limit changes to the service library to fixing the memory issue by creating a separate PR and move the generic UploadTask to Core in a short while so as to support both v1.0 and the beta libs.

**Pull request closed without merging**